src/coreclr/jit/emit.cpp

```diff
@@ -8720,6 +8764,16 @@ void emitter::emitUpdateLiveGCvars(VARSET_VALARG_TP vars, BYTE* addr)
     emitThisGCrefVset = true;
 }
 
+/*****************************************************************************
+ *
+ *  Last emitted instruction is a call and it is not a NoGC call.
+ */
+
+bool emitter::emitLastInsIsCallWithGC()
+{
+    return emitLastIns != nullptr && emitLastIns->idIsCall() && !emitLastIns->idIsNoGC();
+}
+
```

Comment:
We have functions like `emitRemoveLastInstruction()` that makes me a little uncomfortable with this use of `emitLastIns` for correctness purposes.

We may need to look into enabling `EMIT_BACKWARDS_NAVIGATION` more broadly to switch to use it instead. Today we only have this enabled for x64.

Reply:
I believe `emitRemoveLastInstruction` is only used for replacing instructions. 
It does temporarily null out `emitLastIns`, but then always puts something back. 

For example there are peephole optimizations that fuse operations - (as a simplified example: if we have emitted `move x->y` and then we need to emit `move y -> z`, we could instead remove `move x->y` and just put a "fused" `move x->z` instead) 

I do not think there is a scenario where we outright remove an already emitted instruction. If an instruction is redundant (say `move x -> x`), we do not emit it. We will not have any additional info after emitting that we did not before emitting, so that we could go and undo it.

I also do not think an instruction fusing optimization can cross into the next block that has BBF_HAS_LABEL.

Reply:
> I wonder if it would be possible to detect the cases where the NOP/BRK need to be inserted in the forward direction instead?

"Is the next instruction targeted by a branch?" seems harder to answer than "was the last instruction a call?" 

One reason is that some trees may end up not emitting anything. The "what is next" would need to know about all that - what nodes may not emit anything, the cases when we may skip redundant moves, etc...

Reply:
Ok, let's leave it like this then.